File: Heap_Sort/C++/Sweet-kid/heap_sort.cpp
#include <iostream>
using namespace std;

#define SIZE 10

void print_array( int *array, int size );
int parent( int i );
int left( int i );
int right( int i );
void max_heapify( int * array, int i, int size );
void build_max_heap( int * array, int size );
void heap_sort( int * array, int size );
void max_heapify_iterative( int * array, int i, int size );
int test_sorted_array( int * array, int size );

int main() {
  int array[] = { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1 };

  cout << "Unsorted array : " << endl;
  print_array( array, SIZE );

  heap_sort( array, SIZE );

  cout << "Sorted array : " << endl;
  print_array( array, SIZE );

  cout << "Testing if array is actually sorted or not..." << endl;
  if( test_sorted_array( array, SIZE ) ) {
    cout << "Array is sorted." << endl;
  } else {
    cout << "Array is not sorted." << endl;
  }
}

void print_array( int * array, int size ) {
  for( int i = 0; i < size; i++ ) {
    cout << array[ i ] << " ";
  }

  cout << endl << endl;
}

int parent( int i ) {
  return (int)( ( i + 1 )/ 2 ) - 1;
}

int left( int i ) {
  return i * 2 + 1;
}

int right( int i ) {
  return 2 * i + 2;
}

void max_heapify( int * array, int i, int size ) {
  int l = left( i );
  int r = right( i );
  int largest, temp;

  if( l <= (size - 1) && array[ l ] > array[ i ] ) {
    largest = l;
  } else {
    largest = i;
  }

  if( r <= (size - 1) && array[ r ] > array[ largest ] ) {
    largest = r;
  }

  if( largest != i ) {
    temp = array[ i ];
    array[ i ] = array[ largest ];
    array[ largest ] = temp;
    max_heapify( array, largest, size );
  }
}

void max_heapify_iterative( int * array, int i, int size ) {
  int l, r, largest, temp;

  while( 1 ) {
    l = left( i );
    r = right( i );

    if( l <= (size - 1) && array[ l ] > array[ i ] ) {
      largest = l;
    } else {
      largest = i;
    }

    if( r <= (size - 1) && array[ r ] > array[ largest ] ) {
      largest = r;
    }

    if( largest != i ) {
      temp = array[ i ];
      array[ i ] = array[ largest ];
      array[ largest ] = temp;
      i = largest;
    } else {
      break;
    }
  }
}

void build_max_heap(int * array, int size) {
  int i;
  for( i = (int)( size / 2 ) - 1; i >= 0; i-- ) {
    max_heapify( array, i, size );
  }
}

void heap_sort( int * array, int size ) {
  int i, temp, heap_size = size;
  build_max_heap( array, size );
  for( i = size - 1; i > 0; i-- ) {
    temp = array[ i ];
    array[ i ] = array[ 0 ];
    array[ 0 ] = temp;
    max_heapify( array, 0, --heap_size);
  }
}

int test_sorted_array( int * array, int size ) {
  int i;

  for( i = 0; i < size - 2; i++ ) {
    if( array[ i ] > array[ i + 1 ] ) {
      return 0;
    }
  }

  return 1;
}

File: Heap_Sort/Erlang/Sweet-kid/heap_sort.erl
-module(heap_sort).
-export([heapSort/1, heapSort_test/0]).

heapSort( List ) ->
    NewList = buildMaxHeap( List ),
    heapSort( NewList, length( NewList ) ).

heapSort( List, 1 ) ->
    List;

heapSort( List, Size ) ->
    NewList = swapElements( List, 1, Size ),
    NewList2 = maxHeapify( NewList, 1, Size - 1 ),
    heapSort( NewList2, Size - 1 ).

buildMaxHeap( List ) ->
    buildMaxHeap( List, trunc( length( List ) / 2 ) ).

buildMaxHeap( List, I ) ->
    case (I =< length( List ) / 2) and (I > 0 ) of
	true ->
	    NewList = maxHeapify( List, I, length( List ) ),
	    buildMaxHeap( NewList, I - 1 );
	false ->
	    List
    end.

maxHeapify( List, I, Size ) ->
    L = left( I ),
    R = right( I ),

    IElem = lists:nth(I, List ),

    case ( L =< Size ) andalso ( List =/= [] ) andalso ( lists:nth( L, List ) > IElem ) of
	true  -> Largest = L;
	false -> Largest = I
    end,
    LargestElem = lists:nth( Largest, List ),

    case ( R =< Size ) andalso ( List =/= [] ) andalso ( lists:nth( R, List ) > LargestElem ) of
	true  -> FinalLargest = R;
	false -> FinalLargest = Largest
    end,

    if FinalLargest =/= I ->
	    NewList = swapElements( List, I, FinalLargest ),
	    maxHeapify( NewList, FinalLargest, Size );
       true -> List
    end.

left( I ) ->
    2 * I.

right( I ) ->
    2 * I + 1.

swapElements( List, I, J ) ->
    if I < J ->
	    { SubListTillI, SubListAfterI } = lists:split( I, List ),
	    { SubListBetweenIAndJ, SubListAfterJ } = lists:split( J - I, SubListAfterI ),
	    getList( List, J, I, SubListTillI, SubListBetweenIAndJ, SubListAfterJ );
       true ->
	    { SubListTillJ, SubListAfterJ } = lists:split( J, List ),
	    { SubListBetweenJAndI, SubListAfterI } = lists:split( I - J, SubListAfterJ ),
	    getList( List, I, J, SubListTillJ, SubListBetweenJAndI, SubListAfterI )
    end.

getList( List, I, J, SubList1, SubList2, SubList3 ) ->
    L1 = lists:sublist( SubList1, 1, length( SubList1 ) - 1 ),
    L2 = [ lists:nth( I, List ) ],
    L3 = lists:sublist( SubList2, 1, length( SubList2 ) - 1 ),
    L4 = [ lists:nth( J, List ) ],
    L5 = lists:sublist( SubList3, 1, length( SubList3 ) ),
    L1 ++ L2 ++ L3 ++ L4 ++ L5.

heapSort_test() ->
    [ 1, 2, 3, 4, 5 ] = heapSort( [ 3, 5, 1, 2, 4 ] ),
    [ 6, 12, 34, 45, 67, 89, 90, 96 ] = heapSort( [ 12, 34, 45, 90, 67, 96, 89, 6 ] ).

